libtenzir/builtins/operators/grok.cpp

```diff
@@ -0,0 +1,518 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/argument_parser.hpp>
+#include <tenzir/arrow_table_slice.hpp>
+#include <tenzir/plugin.hpp>
+#include <tenzir/series_builder.hpp>
+
+// Both Boost.Regex and RE2 are used:
+//  - Boost.Regex is used for actual grokking
+//  - RE2 is used for parsing the patterns we're given
+//
+// RE2 can't be used for both, because it doesn't support all the regex features
+// the built-in patterns need.
+//
+// Boost.Regex _could_ be used for both, but it's slow, so we're using RE2 where
+// we can.
+#include <boost/regex.hpp>
+#include <caf/make_copy_on_write.hpp>
+#include <re2/re2.h>
+
+#include <algorithm>
+#include <ranges>
+#include <span>
+
+namespace caf {
+template <>
+struct inspector_access<boost::regex> : inspector_access_base<boost::regex> {
+  template <typename Inspector>
+  static auto apply(Inspector& f, boost::regex& x) {
+    auto str = x.str();
+    auto result = f.apply(str);
+    if constexpr (Inspector::is_loading)
+      x.assign(str);
+    return result;
+  }
+};
+} // namespace caf
+
+namespace tenzir::plugins::grok {
+
+namespace {
+
+constexpr std::string_view builtin_patterns_strings[] =
+#include "./grok-patterns/patterns.inc"
+  ;
+
+enum class capture_type {
+  // The three options below (string, int, float)
+  // are grok standard options
+
+  // %{::string}, default
+  string,
+  // %{::int}
+  integer,
+  // %{::float}
+  floating,
+
+  // Replacement fields without a NAME, only a SYNTAX
+  unnamed,
+
+  // "implicit" named regex capture group,
+  // without %{...},
+  // but with (?<NAME>...) or (?'NAME'...)
+  // Used as a tag: in terms of type conversion, this is the same as `string`
+  implicit,
+};
```

Comment:
I don't know. I intentionally did no modifications to the built-in patterns, to ensure portability, and to prevent myself from going down an endless rabbit hole of minute adjustments. I also don't think we should be special-casing any built-in patterns.

Possible solutions:
 1. Add an option called `infer` (or `auto`, or something else), which could try parsing a string, and see if it could be represented better as another type. This could maybe also be made the default, which would be a breaking change compared to Logstash grok, which defaults to `string`. It should also be careful: the text "null" should probably not be treated as a null value.
 2. Add an option for (some subset of) our own types, in addition to what's already there, so `bool`, `int64`, `uint64`, `double`, `duration`, `time`, `ip`, `subnet`. (I skipped `none`, `pattern`, `list` and `record`).
 3. Do both?

With (1), `%{IP:addr}` would output an `ip` if we made `infer` the default, otherwise `%{IP:addr:infer}` would do the trick. With (2), `%{IP:addr}` would keep producing a string, and `%{IP:addr:ip}` would yield an `ip`.

Reply:
It's also somewhat difficult to programmatically ascertain what type the pattern would map to. For example, the pattern `%{DATE_EU}` (defined as `%{MONTHDAY}[./-]%{MONTHNUM}[./-]%{YEAR}`) clearly _means_ a date, but without additionally telling the parser that it's a date in the format `DD(.|/|-)MM(.|/|-)YYYY`, there's no way to parse it reliably.

Reply:
This is why the optimal solution would probably be to only do limited type inference/casting in `grok`, and leave the rest for other operators. Our revamped language is starting to sound more grandiose by the second.

```
# extremely hypothetical
... | parse grok --include-unnamed "%{DATE_EU}" | parse timestamp f"{self.YEAR}-{self.MONTHNUM}-{self.MONTHDAY}"
```

Reply:
Let's leave this to a potential follow-up PR.